rust/frontend/src/ndarray.rs

```diff
@@ -0,0 +1,359 @@
+    /// let ctx = TVMContext::cpu(0);
+    /// let mut ndarray = empty(&mut shape, ctx, TVMType::from("int"));
+    /// ndarray.copy_from_buffer(&mut data);
+    /// assert_eq!(ndarray.shape(), Some(shape));
+    /// assert_eq!(ndarray.to_vec::<i32>().unwrap(), data);
+    /// ```
+    pub fn to_vec<T>(&self) -> Result<Vec<T>> {
+        if self.shape().is_none() {
+            return Err(Error::EmptyArray);
+        }
+        let earr = empty(&mut self.shape().unwrap(), TVMContext::cpu(0), self.dtype());
+        let target = self.copy_to_ndarray(earr).unwrap();
+        let arr = unsafe { *(target.handle) };
+        let sz = self.size().unwrap() as usize;
+        let mut v: Vec<T> = Vec::with_capacity(sz * mem::size_of::<T>());
+        unsafe {
+            v.as_mut_ptr()
+                .copy_from_nonoverlapping(arr.data as *const T, sz);
+            v.set_len(sz);
+        }
+        Ok(v)
+    }
+
+    /// Converts the NDArray to [`TVMByteArray`].
+    pub fn to_bytearray(&self) -> Result<TVMByteArray> {
+        let v = self.to_vec::<u8>()?;
+        Ok(TVMByteArray::from(&v))
+    }
+
+    /// Creates an NDArray from a mutable buffer of types i32, u32 or f32 in cpu.
+    /// ## Example
+    ///
+    /// ```
+    /// let mut shape = vec![2];
+    /// let mut data = vec![1f32, 2];
+    /// let ctx = TVMContext::gpu(0);
+    /// let mut ndarray = empty(&mut shape, ctx, TVMType::from("int"));
+    /// ndarray.copy_from_buffer(&mut data);
+    /// ```
+    pub fn copy_from_buffer<T: Num32>(&mut self, data: &mut [T]) {
+        check_call!(ts::TVMArrayCopyFromBytes(
+            self.handle,
+            data.as_ptr() as *mut _,
+            data.len() * mem::size_of::<T>()
+        ));
+    }
+
+    /// Copies the NDArray to another target NDArray.
+    pub fn copy_to_ndarray(&self, target: NDArray) -> Result<NDArray> {
+        if self.dtype() != target.dtype() {
+            return Err(Error::TypeMismatch {
+                expected: self.dtype().to_string(),
+                found: target.dtype().to_string(),
+            });
+        }
+        check_call!(ts::TVMArrayCopyFromTo(
+            self.handle,
+            target.handle,
+            ptr::null_mut() as ts::TVMStreamHandle
+        ));
+        Ok(target)
+    }
+
+    /// Copies the NDArray to a target context.
+    pub fn copy_to_ctx(&self, target: &TVMContext) -> Result<NDArray> {
+        let tmp = empty(&mut self.shape().unwrap(), target.clone(), self.dtype());
+        let copy = self.copy_to_ndarray(tmp)?;
+        Ok(copy)
+    }
+
+    /// Converts a Rust's ndarray to TVM NDArray.
+    pub fn from_rust_ndarray<T: Num32 + Copy>(
+        rnd: &ArrayD<T>,
+        ctx: TVMContext,
+        dtype: TVMType,
+    ) -> Result<Self> {
+        let mut shape = rnd.shape().to_vec();
+        let mut nd = empty(&mut shape, ctx, dtype);
+        let mut vec = Array::from_iter(rnd.iter())
+            .iter()
+            .map(|e| **e)
+            .collect::<Vec<T>>();
```

Comment:
double copy. you should directly convert `rnd`

Reply:
Well, after I did the above the result of resnet for a cat change into window! It clearly messes up the order ndarray entries. I need a beer now!

Reply:
ah, the trick is that you still do actually need to convert to `<T>`. I had only meant to say that you can do something closer to

```
Array::from_iter(rnd.into_iter().map(|v| v as T))
```